# Pregunta 4

## Con el uso de librerías realiza en Python los mismos preprocesamiento del punto 3.

### R.-

Primero, como los colores son valores **continuos**, primero normalizaremos y discretizaremos el dataset. Como cada dataset tiene 4096 columnas, la **normalización** y **dicretización** serán **binarias**, así cuando hagamos el `onehotencoder`, y el `labelencoder` las columnas serán 8192 por dataset.

In [30]:
import os

current_dir = os.getcwd()

input_folder = os.path.join(os.path.dirname(current_dir), "Pregunta 2/dataset02_2")


In [51]:
import os
import numpy as np
import pandas as pd
from PIL import Image
from sklearn.preprocessing import MinMaxScaler

red_data_unnormalized = []
green_data_unnormalized = []
blue_data_unnormalized = []
gray_data_unnormalized = []

# Inicializa listas para almacenar los datos de cada canal
red_data = []
green_data = []
blue_data = []
gray_data = []

# Recorre cada imagen en el directorio
for filename in os.listdir(input_folder):
    if filename.endswith('.jpg'):
        img_path = os.path.join(input_folder, filename)
        
        # Carga la imagen
        img = Image.open(img_path)
        
        # Convierte la imagen a RGB
        img = img.convert('RGB')
        
        # Convierte la imagen a un array de numpy
        img_array = np.array(img)
        
        # Separa los canales rojo, verde y azul
        red_channel = img_array[:,:,0].flatten()   # 4096 elementos
        green_channel = img_array[:,:,1].flatten() # 4096 elementos
        blue_channel = img_array[:,:,2].flatten()  # 4096 elementos
        
        # Calcula la escala de grises
        grayscale = np.dot(img_array[...,:3], [0.2989, 0.5870, 0.1140])  # Convierte a escala de grises
        gray_channel = grayscale.flatten()  # 4096 elementos

        # Normalizar y discretizar entre 0 y 1
        scaler = MinMaxScaler()
        
        red_data_unnormalized.append(red_channel)
        green_data_unnormalized.append(green_channel)
        blue_data_unnormalized.append(blue_channel)
        gray_data_unnormalized.append(gray_channel)
        
        
        red_channel = red_channel.reshape(-1,1)
        green_channel = green_channel.reshape(-1,1)
        blue_channel = blue_channel.reshape(-1,1)
        gray_channel = gray_channel.reshape(-1,1)        

        red_channel = scaler.fit_transform(red_channel)
        green_channel = scaler.fit_transform(green_channel)
        blue_channel = scaler.fit_transform(blue_channel)
        gray_channel = scaler.fit_transform(gray_channel)
                
        red_channel = (red_channel >= 0.5).astype(int)
        green_channel = (green_channel >= 0.5).astype(int)
        blue_channel = (blue_channel >= 0.5).astype(int)
        gray_channel = (gray_channel >= 0.5).astype(int)   

        # Agrega los canales a sus respectivas listas
        red_data.append(red_channel.flatten())
        green_data.append(green_channel.flatten())
        blue_data.append(blue_channel.flatten())
        gray_data.append(gray_channel.flatten())

In [ ]:
dataFramesUnnormalized=[]

for data in [red_data_unnormalized, green_data_unnormalized, blue_data_unnormalized, gray_data_unnormalized]:
    df = pd.DataFrame(data)  # Cada lista tiene 5000 elementos, cada uno con 4096 valores
    dataFramesUnnormalized.append(df)

Coleccionando dataframes

In [40]:
dataFrames=[]

for data in [red_data, green_data, blue_data, gray_data]:
    df = pd.DataFrame(data)  # Cada lista tiene 1000 elementos, cada uno con 4096 valores
    dataFrames.append(df)    

    # df.to_csv(tempName, index=False, header=False)

In [53]:
dataFramesUnnormalized=[]

for data in [red_data_unnormalized, green_data_unnormalized, blue_data_unnormalized, gray_data_unnormalized]:
    df = pd.DataFrame(data)  # Cada lista tiene 1000 elementos, cada uno con 4096 valores
    dataFramesUnnormalized.append(df)

Cada imagen tiene $64x64$ pixeles, oses $4096$, y son $5000$ imagenes.

In [42]:
print(dataFrames[0])

      0     1     2     3     4     5     6     7     8     9     ...  4086  \
0        1     1     1     1     1     1     0     0     1     0  ...     1   
1        0     1     0     1     1     1     0     0     0     0  ...     1   
2        1     1     1     1     1     1     1     1     1     0  ...     0   
3        0     0     0     0     0     1     1     1     0     0  ...     0   
4        1     1     1     1     1     1     1     1     1     1  ...     0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4995     0     0     0     0     0     0     0     0     0     0  ...     1   
4996     1     1     1     1     1     1     1     1     1     1  ...     1   
4997     1     1     1     1     1     1     1     1     1     1  ...     0   
4998     1     1     1     1     1     1     1     1     1     1  ...     0   
4999     0     1     1     1     1     1     1     1     1     1  ...     1   

      4087  4088  4089  4090  4091  4092  4093  409

OneHotEncoder

In [45]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder

dataFramesOnHot = []

for i in range(4):
    df = dataFrames[i]

    encoder = OneHotEncoder(sparse_output=False)  # sparse_output=False para obtener un array denso

    # Ajustar y transformar el DataFrame
    one_hot_encoded = encoder.fit_transform(df)

    # Convertir el resultado a enteros
    one_hot_encoded_int = one_hot_encoded.astype(int)

    # Convertir a DataFrame para verlo mejor
    df_one_hot = pd.DataFrame(one_hot_encoded_int)

    dataFramesOnHot.append(df_one_hot)
    
print(dataFramesOnHot[0])

      0     1     2     3     4     5     6     7     8     9     ...  8182  \
0        0     1     0     1     0     1     0     1     0     1  ...     0   
1        1     0     0     1     1     0     0     1     0     1  ...     0   
2        0     1     0     1     0     1     0     1     0     1  ...     1   
3        1     0     1     0     1     0     1     0     1     0  ...     1   
4        0     1     0     1     0     1     0     1     0     1  ...     1   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4995     1     0     1     0     1     0     1     0     1     0  ...     0   
4996     0     1     0     1     0     1     0     1     0     1  ...     0   
4997     0     1     0     1     0     1     0     1     0     1  ...     1   
4998     0     1     0     1     0     1     0     1     0     1  ...     1   
4999     1     0     0     1     0     1     0     1     0     1  ...     0   

      8183  8184  8185  8186  8187  8188  8189  819

En efecto, las columnas de doblaron, ahora son 8182

In [58]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

# Función para clasificar intensidades en categorías
def classify_intensity(value):
    if value < 85:
        return 'L' # Low
    elif value < 170:
        return 'M' # Medium
    else:
        return 'H' # High

# Aplica la clasificación a cada DataFrame de canal
red_channel_categorized = dataFramesUnnormalized[0].applymap(classify_intensity)
green_channel_categorized = dataFramesUnnormalized[1].applymap(classify_intensity)
blue_channel_categorized = dataFramesUnnormalized[2].applymap(classify_intensity)
gray_channel_categorized = dataFramesUnnormalized[3].applymap(classify_intensity)


# Usar LabelEncoder para transformar las categorías
encoder = LabelEncoder()

# Transforma cada canal
red_encoded = red_channel_categorized.apply(encoder.fit_transform)
green_encoded = green_channel_categorized.apply(encoder.fit_transform)
blue_encoded = blue_channel_categorized.apply(encoder.fit_transform)
gray_encoded = gray_channel_categorized.apply(encoder.fit_transform)


/tmp/ipykernel_7439/1668296745.py:15: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  red_channel_categorized = dataFramesUnnormalized[0].applymap(classify_intensity)


**LabelEncoder** con etiquetas textuales

In [57]:
print(red_channel_categorized)

     0    1    2    3    4    5    6    7    8    9     ... 4086 4087 4088  \
0       H    H    H    H    H    H    M    L    M    L  ...    H    H    H   
1       M    M    M    M    M    M    M    M    M    M  ...    H    H    M   
2       H    H    H    H    H    H    M    H    H    L  ...    M    M    M   
3       M    M    M    M    M    M    M    M    M    M  ...    L    L    L   
4       H    H    H    H    H    H    H    H    H    H  ...    L    M    M   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
4995    L    L    L    L    L    L    L    L    L    L  ...    H    H    H   
4996    M    M    M    M    M    M    M    M    M    M  ...    M    M    M   
4997    H    H    H    H    H    H    H    H    H    H  ...    M    M    M   
4998    M    H    H    H    M    H    H    H    H    H  ...    M    M    M   
4999    M    M    M    M    M    H    H    H    H    H  ...    M    M    M   

     4089 4090 4091 4092 4093 4094 4095  
0       H    H    H  

**LabelEncoder** con etiquetas numéricas

In [59]:
print(red_encoded)

      0     1     2     3     4     5     6     7     8     9     ...  4086  \
0        0     0     0     0     0     0     2     1     2     1  ...     0   
1        2     2     2     2     2     2     2     2     2     2  ...     0   
2        0     0     0     0     0     0     2     0     0     1  ...     2   
3        2     2     2     2     2     2     2     2     2     2  ...     1   
4        0     0     0     0     0     0     0     0     0     0  ...     1   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4995     1     1     1     1     1     1     1     1     1     1  ...     0   
4996     2     2     2     2     2     2     2     2     2     2  ...     2   
4997     0     0     0     0     0     0     0     0     0     0  ...     2   
4998     2     0     0     0     2     0     0     0     0     0  ...     2   
4999     2     2     2     2     2     0     0     0     0     0  ...     2   

      4087  4088  4089  4090  4091  4092  4093  409